### PR TITLE
[SQL] Sql midb Add: list-deleted, show-deleted, update-retention, show-retention

### PR DESCRIPTION
**Pull request by djnisic**
Adding new cmdlets that will support short term retention and restore of deleted databases on MI.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
add to S165.

**Comment by jaredmoo** [comment]
For `az sql db` I intentionally implemented only `list-deleted` and not `show-deleted` because the resource name of deleted db is a bit weird and getting single deleted db seems like a relatively corner case scenario. What are your thoughts?

File: src/azure-cli/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py [hunk 1]
```diff
@@ -2954,6 +2954,214 @@ def test_sql_mi_tdebyok(self, resource_group, resource_group_location):
...
+
+        # test get short term retention on live database
+        self.cmd('sql midb show-retention -g {rg} --mi {managed_instance_name} -n {database_name}',
+                 checks=[
+                     self.check('resourceGroup', '{rg}'),
+                     self.check('retentionDays', '{retention_days_inc}')])
+
+        # Wait for first backup before dropping
+        if self.in_recording:
+            time.sleep(300)
```
Comment (adeal) [inline src/azure-cli/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py]: we should change this to polling to reduce chances of flakiness...you can query the GetDatabase API until EasrliestRestoreDate becomes available. Thoughts?

Comment (jaredmoo) [reply src/azure-cli/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py]: Agreed, the existing test for sql db pitr will wait for earliest restore date.

Comment (djnisic) [reply src/azure-cli/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py]: Sure, I will add that.

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 2]
```diff
@@ -596,12 +596,52 @@
     text: az sql midb list -g mygroup --mi myinstance
 """
 
+helps['sql midb list-deleted'] = """
+type: command
+short-summary: Lists restorable deleted managed databases.
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py]: Lists -> List

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 3]
```diff
@@ -596,12 +596,52 @@
...
+examples:
+  - name: List all restorable deleted managed databases on Managed Instance .
+    text: az sql midb list-deleted -g mygroup --mi myinstance
+"""
+
+helps['sql midb show-deleted'] = """
+type: command
+short-summary: Shows restorable deleted managed database.
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py]: Shows -> Show

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 4]
```diff
@@ -596,12 +596,52 @@
...
     text: az sql midb restore -g mygroup --mi myinstance -n mymanageddb --dest-name targetmidb --time "2018-05-20T05:34:22"
+  - name: Restore a dropped managed database using Point in time restore
+    text: az sql midb restore -g mygroup --mi myinstance -n mymanageddb --dest-name targetmidb --time "2018-05-20T05:34:22" --deletion-date "2018-05-20T05:34:22"
+  - name: Restore a live managed database using Point in time restore from another instance
+    text: az sql midb restore -g mygroup --mi myinstance -n mymanageddb --dest-name targetmidb --time "2018-05-20T05:34:22" --dest-mi targetmi --dest-resource-group targetrg
+"""
+
+helps['sql midb update-retention'] = """
+type: command
+short-summary: Updates short term retention for automated backups on a single database.
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py]: Updates -> Update

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 5]
```diff
@@ -596,12 +596,52 @@
...
+examples:
+  - name: Update backup short term retention for live managed database.
+    text: az sql midb update-retention -g mygroup --mi myinstance -n mymanageddb --retention-days retentionindays
+  - name: Update backup short term retention for dropped managed database.
+    text: az sql midb update-retention -g mygroup --mi myinstance -n mymanageddb --deletion-date "2018-05-20T05:34:22" --retention-days retentionindays
+"""
+
+helps['sql midb show-retention'] = """
+type: command
+short-summary: Shows short term retention for automated backups on a single database.
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py]: Shows -> Show

File: src/azure-cli/azure/cli/command_modules/sql/commands.py [hunk 6]
```diff
@@ -220,6 +222,17 @@ def load_command_table(self, _):
...
+    restorable_dropped_managed_databases_operations = CliCommandType(
+        operations_tmpl='azure.mgmt.sql.operations#RestorableDroppedManagedDatabasesOperations.{}',
+        client_factory=get_sql_restorable_dropped_managed_databases_operations)
+
+    with self.command_group('sql midb',
+                            restorable_dropped_managed_databases_operations,
+                            client_factory=get_sql_restorable_dropped_managed_databases_operations) as g:
+
+        g.command('list-deleted', 'list_by_instance')
+        g.custom_command('show-deleted', 'show_deleted_database_mi')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/commands.py]: The two commands are a little werid. What about add a new parameter `--include-deleted` in existing show and list commands like storage blob command?

File: src/azure-cli/azure/cli/command_modules/sql/commands.py [hunk 7]
```diff
@@ -547,6 +560,17 @@ def load_command_table(self, _):
...
+    managed_backup_short_term_retention_policies_operations = CliCommandType(
+        operations_tmpl='azure.mgmt.sql.operations#ManagedBackupShortTermRetentionPoliciesOperations.{}',
+        client_factory=get_sql_managed_backup_short_term_retention_policies_operations)
+
+    with self.command_group('sql midb',
+                            managed_backup_short_term_retention_policies_operations,
+                            client_factory=get_sql_managed_backup_short_term_retention_policies_operations) as g:
+
+        g.custom_command('update-retention', 'update_short_term_retention_mi', supports_no_wait=True)
+        g.custom_command('show-retention', 'get_short_term_retention_mi')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/commands.py]: What about `az sql midb delete-retention update/show` or `az sql midb delete-policy update/show` ?

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 8]
```diff
@@ -457,6 +474,45 @@ def _get_managed_db_resource_id(cli_ctx, resource_group_name, managed_instance_n
...
+        deletion_date):
+    '''
+    Gets the Managed db resource id in this Azure environment.
+    '''
+
+    # url parse package has different names in Python 2 and 3. 'six' package works cross-version.
+    from six.moves.urllib.parse import quote  # pylint: disable=import-error
+    from azure.cli.core.commands.client_factory import get_subscription_id
+
+    return ('/subscriptions/{}/resourceGroups/{}/providers/Microsoft.Sql/managedInstances/'
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py]: You can consider using resource_id in msrestazure. `from msrestazure.tools import resource_id`

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 9]
```diff
@@ -2498,23 +2555,129 @@ def managed_db_restore(
...
+
+    Custom function makes create mode more convenient.
+    '''
+
+    if deletion_date:
+        database_name = '{},{}'.format(
+            database_name,
+            _to_filetimeutc(deletion_date))
+
+        client = \
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py]: why not use client in parameter list here?

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 10]
```diff
@@ -2498,23 +2555,129 @@ def managed_db_restore(
...
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py]: The same as above.

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 11]
```diff
@@ -2498,23 +2555,129 @@ def managed_db_restore(
...
+            get_sql_restorable_dropped_database_managed_backup_short_term_retention_policies_operations(
+                cmd.cli_ctx,
+                None)
+
+        policy = client.get(
+            restorable_dropped_database_id=database_name,
+            managed_instance_name=managed_instance_name,
+            resource_group_name=resource_group_name)
+    else:
+        policy = client.get(
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py]: If deletion date is given, you will return dropped database related information and if not, you will return database information directly, right?
Can live database and dropped database be the same name? If yes, in your current scenario, as soon as deletion date is given, you will only show up the dropped database information, right?

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 12]
```diff
@@ -596,12 +596,52 @@
...
+examples:
+  - name: List all restorable deleted managed databases on Managed Instance .
+    text: az sql midb list-deleted -g mygroup --mi myinstance
+"""
+
+helps['sql midb show-deleted'] = """
+type: command
+short-summary: Shows restorable deleted managed database.
+examples:
+  - name: Shows specific restorable deleted managed database.
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py]: Shows -> Show

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 13]
```diff
@@ -596,12 +596,52 @@
...
+    text: az sql midb update-retention -g mygroup --mi myinstance -n mymanageddb --deletion-date "2018-05-20T05:34:22" --retention-days retentionindays
+"""
+
+helps['sql midb show-retention'] = """
+type: command
+short-summary: Shows short term retention for automated backups on a single database.
+examples:
+  - name: Shows backup short term retention for live managed database.
+    text: az sql midb show-retention -g mygroup --mi myinstance -n mymanageddb
+  - name: Shows backup short term retention for dropped managed database.
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py]: Shows -> Show

**Comment by Juliehzl** [comment]
Another suggestion here is that can we make the new commands in Preview flag?

File: src/azure-cli/azure/cli/command_modules/sql/commands.py (hunk 6 above)
Comment (djnisic) [reply src/azure-cli/azure/cli/command_modules/sql/commands.py]: That could work, however I wanted to keep cmdlets for MI similar to the SQLDB ones. As you can see couple of lines above SQLDB is using list-deleted. 

Other reason is that live and dropped databases are not same resource type. In PowerShell that breaks piping but I am not sure if this is an issue in CLI.

File: src/azure-cli/azure/cli/command_modules/sql/custom.py (hunk 11 above)
Comment (djnisic) [reply src/azure-cli/azure/cli/command_modules/sql/custom.py]: Correct, there can exist live and multiple dropped databases with the same name. If deletion date is provided I will use client for dropped databases and fetch them, otherwise I will use client for live databases.

**Comment by djnisic** [comment]
> For `az sql db` I intentionally implemented only `list-deleted` and not `show-deleted` because the resource name of deleted db is a bit weird and getting single deleted db seems like a relatively corner case scenario. What are your thoughts?

I think you are right. In order to use show for dropped databases, you would have to use list first and get it's dropped time, which defeats it's purpose.

File: src/azure-cli/azure/cli/command_modules/sql/commands.py (hunk 7 above)
Comment (djnisic) [reply src/azure-cli/azure/cli/command_modules/sql/commands.py]: You mean as in renaming current cmdlets or as in adding a new ones for deleted databases. 

If we are talking about altering existing, we are open for suggestions. az sql midb retention-policy update/show could also be an option. However I would avoid delete-retention as this retention is for automated backups on database not database itself.

I wouldn't separate cmdlets for dropped and live databases since policy doesn't change once database is dropped. We might separate them in the future though.

File: src/azure-cli/azure/cli/command_modules/sql/commands.py (hunk 6 above)
Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/sql/commands.py]: I see. The reason I recommend such design is that we have similar implementation in storage commands and what we do is add an new parameters as what I have mentioned. And as far as I know, Powershell also has similar design. 
But as you said, in your scenario, it will break piping in Powershell. So there maybe some difference.
In this way, I am fine with the command if you already have similar design.

File: src/azure-cli/azure/cli/command_modules/sql/commands.py (hunk 7 above)
Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/sql/commands.py]: I mean rename currrent cmdlets. You can find the option you like. 

Actually, the reason I have such suggestion is that cli do not recommend commands with hyphen for the last part. The only scenario to use hyphen in command is that the abstracted subgroup only has single command, e.g.  `az stroage account show-connection-string`.

**Comment by Juliehzl** [comment]
No other concern to me. @djnisic please resolve my comments ASAP if you want to catch the next realase tomorrow and make sure @jaredmoo approve.

**Comment by djnisic** [comment]
> No other concern to me. @djnisic please resolve my comments ASAP if you want to catch the next realase tomorrow and make sure @jaredmoo approve.

Will do. I am currently running tests with Austin's suggestion. Each tests takes a lot for MI(~4 hours). Regarding new release, I don't think this is that critical for MI, but I will try to get it done today.

**Comment by djnisic** [comment]
@jaredmoo everything should be OK now, please review.

File: src/azure-cli/azure/cli/command_modules/sql/_params.py [hunk 14]
```diff
@@ -1419,11 +1419,17 @@ def _configure_security_policy_storage_params(arg_ctx):
...
         create_args_for_complex_type(
             c, 'parameters', ManagedDatabase, [
+                'deletion_date',
                 'target_managed_database_name',
                 'target_managed_instance_name',
                 'restore_point_in_time'
             ])
 
+        c.argument('deletion_date',
+                   options_list=['--deletion-date'],
```
Comment (jaredmoo) [inline src/azure-cli/azure/cli/command_modules/sql/_params.py]: Can you align this with `az sql db restore`, which has `--deleted-time`? 

https://docs.microsoft.com/en-us/cli/azure/sql/db?view=azure-cli-latest#az-sql-db-restore

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 15]
```diff
@@ -67,6 +72,18 @@ def _get_server_location(cli_ctx, server_name, resource_group_name):
         resource_group_name=resource_group_name).location
 
 
+def _get_managed_restorable_dropped_database_backup_short_term_retention_client(cli_ctx):
+    '''
+    Returns the location (i.e. Azure region) that the specified server is in.
```
Comment (jaredmoo) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py]: Is this comment correct?

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 16]
```diff
@@ -446,15 +463,59 @@ def _get_managed_db_resource_id(cli_ctx, resource_group_name, managed_instance_n
...
+        subscription=get_subscription_id(cli_ctx),
+        resource_group=resource_group_name,
+        namespace='Microsoft.Sql', type='managedInstances',
+        name=managed_instance_name,
+        child_type_1='databases',
+        child_name_1=database_name)
+
+
+def _to_filetimeutc(dateTime):
+    NET_epoch = datetime(1601, 1, 1)
```
Comment (jaredmoo) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py]: Add function comment

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 17]
```diff
@@ -2509,23 +2571,129 @@ def managed_db_restore(
...
+
+
+def show_deleted_database_mi(
+        client,
+        database_name,
+        managed_instance_name,
+        resource_group_name,
+        deletion_date):
+    '''
+    Updates short term retention for database
```
Comment (jaredmoo) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py]: This comment is also not correct :)

**Comment by jaredmoo** [comment]
Please also make sure that @adeal  approves. Cheers

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 18]
```diff
@@ -601,12 +601,24 @@
...
+
 helps['sql midb restore'] = """
 type: command
 short-summary: Restore a managed database.
 examples:
-  - name: Restore a managed database using Point in time restore
+  - name: Restore a live managed database using Point in time restore
     text: az sql midb restore -g mygroup --mi myinstance -n mymanageddb --dest-name targetmidb --time "2018-05-20T05:34:22"
+  - name: Restore a dropped managed database using Point in time restore
+    text: az sql midb restore -g mygroup --mi myinstance -n mymanageddb --dest-name targetmidb --time "2018-05-20T05:34:22" --deleted-time "2018-05-20T05:34:22"
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py]: What is the difference for --time and --deleted-time?

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 19]
```diff
@@ -617,6 +629,31 @@
...
 """
 
+helps['sql midb retention-policy'] = """
+type: group
+short-summary: Manage SQL managed instance database backup retention policy.
+"""
+
+helps['sql midb retention-policy update'] = """
+type: command
+short-summary: Update short term retention for automated backups on a single database.
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py]: Hi, I am trying to understand your service.  And I am a little confused about the retention policy here. Can you help me better understand your service? I am not sure if your service also has delete retention policy and restore policy. And for your retention policy days, is it defined to specify the days how long the deleted sql database will be retained or restored?

File: src/azure-cli/azure/cli/command_modules/sql/_params.py [hunk 20]
```diff
@@ -1450,8 +1456,28 @@ def _configure_security_policy_storage_params(arg_ctx):
...
+
+        c.argument('deleted_time',
+                   options_list=['--deleted-time'],
+                   help='If specified, updates retention days for a deleted database, instead of an existing database.'
+                   'Must match the deleted time of a deleted database on the source Managed Instance.')
+
+        c.argument('retention_days',
+                   options_list=['--retention-days'],
+                   required=True,
+                   help='New retention policy in days.')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/sql/_params.py]: Can you give a more descriptive help here? and is there a valid range for the days?

File: src/azure-cli/azure/cli/command_modules/sql/_help.py (hunk 18 above)
Comment (djnisic) [reply src/azure-cli/azure/cli/command_modules/sql/_help.py]: --deleted-time is required to identify dropped database, because primary key to identify dropped database is database name + drop time. For live databases name is good enough, because you cannot have multiple databases with the same name. --time on other had represents the point we are restoring to.

File: src/azure-cli/azure/cli/command_modules/sql/_help.py (hunk 19 above)
Comment (djnisic) [reply src/azure-cli/azure/cli/command_modules/sql/_help.py]: We are taking automated backups for every database on MI. Automated backups can be used to restore databases to any point withing the retention period. This policy will determine for how long are we going to keep those backups before deleting them(retention period). 

There is only one policy per database, and it ranges for 7 to 35 for live databases, and 0 to 35 for dropped databases. Note we are going to reduce minimum retention days for live databases to 1 in the future.

File: src/azure-cli/azure/cli/command_modules/sql/_params.py (hunk 20 above)
Comment (djnisic) [reply src/azure-cli/azure/cli/command_modules/sql/_params.py]: Sure I can add additional information here.

**Comment by djnisic** [comment]
@adeal ping.

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 21]
```diff
@@ -601,12 +601,24 @@
...
 helps['sql midb restore'] = """
 type: command
 short-summary: Restore a managed database.
 examples:
-  - name: Restore a managed database using Point in time restore
+  - name: Restore a live managed database using Point in time restore
     text: az sql midb restore -g mygroup --mi myinstance -n mymanageddb --dest-name targetmidb --time "2018-05-20T05:34:22"
+  - name: Restore a dropped managed database using Point in time restore
+    text: az sql midb restore -g mygroup --mi myinstance -n mymanageddb --dest-name targetmidb --time "2018-05-20T05:34:22" --deleted-time "2018-05-20T05:34:22"
+  - name: Restore a live managed database using Point in time restore from another instance
```
Comment (adeal) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py]: nit: I think it might be more clear to say "Restore a live managed database from another instance using Point in time restore"

File: src/azure-cli/azure/cli/command_modules/sql/_help.py (hunk 19 above)
Comment (adeal) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py]: Other clients (PS) have the following as the documentation for this api, do we want consistency?

> Sets a backup short term retention policy.

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 22]
```diff
@@ -617,6 +629,31 @@
...
+type: command
+short-summary: Update short term retention for automated backups on a single database.
+examples:
+  - name: Update backup short term retention for live managed database.
+    text: az sql midb retention-policy update -g mygroup --mi myinstance -n mymanageddb --retention-days retentionindays
+  - name: Update backup short term retention for dropped managed database.
+    text: az sql midb retention-policy update -g mygroup --mi myinstance -n mymanageddb --deleted-time "2018-05-20T05:34:22" --retention-days retentionindays
+"""
+
+helps['sql midb retention-policy show'] = """
```
Comment (adeal) [inline src/azure-cli/azure/cli/command_modules/sql/_help.py]: I think we need to be explicit in what retention-policy is being updated. Our naming convention is unfortunate, but it's best to make sure the apis and clients all align. From that, we should use short-term-retention-policy

File: src/azure-cli/azure/cli/command_modules/sql/_params.py [hunk 23]
```diff
@@ -1419,11 +1419,17 @@ def _configure_security_policy_storage_params(arg_ctx):
     with self.argument_context('sql midb restore') as c:
         create_args_for_complex_type(
             c, 'parameters', ManagedDatabase, [
+                'deleted_time',
```
Comment (adeal) [inline src/azure-cli/azure/cli/command_modules/sql/_params.py]: `deleted_time` does not align with other apis/clients as well. I think it should be `dropped_time`?

Comment (jaredmoo) [reply src/azure-cli/azure/cli/command_modules/sql/_params.py]: Deleted-time is correct: https://docs.microsoft.com/en-us/cli/azure/sql/db?view=azure-cli-latest#az-sql-db-restore

Consistency with an experience (e.g.: AzCLI, PowerShell, or Portal) is higher priority than consistency across experiences.

Comment (jaredmoo) [reply src/azure-cli/azure/cli/command_modules/sql/_params.py]: Note that I deliberately chose to use `--deleted-time` when implementing `az sql db restore` because the command is `az sql db delete`, not `az sql db drop`.

Comment (adeal) [reply src/azure-cli/azure/cli/command_modules/sql/_params.py]: Makes sense and I agree on consistency throughout any one given client. Thanks for the extra context!

File: src/azure-cli/azure/cli/command_modules/sql/_help.py (hunk 21 above)
Comment (djnisic) [reply src/azure-cli/azure/cli/command_modules/sql/_help.py]: Sure, will change.

File: src/azure-cli/azure/cli/command_modules/sql/_help.py (hunk 19 above)
Comment (djnisic) [reply src/azure-cli/azure/cli/command_modules/sql/_help.py]: You are right, will change update to set.

File: src/azure-cli/azure/cli/command_modules/sql/_help.py (hunk 22 above)
Comment (djnisic) [reply src/azure-cli/azure/cli/command_modules/sql/_help.py]: Sure.

**Comment by djnisic** [comment]
@adeal, @Juliehzl all comments addressed. Can we merge pull request now?

**Comment by djnisic** [comment]
Ping!

**Comment by djnisic** [comment]
@jaredmoo, @Juliehzl ping.
